docs/proposals/shoot-control-plane-migration/README.md

```diff
@@ -0,0 +1,234 @@
+
+To address this issue, the co-ordinator can be supplied an additional source for the current lease status via some resource within the same namespace as the co-ordinator. The co-ordinator can choose to use the lease status from such a local resource or the remote/global/distributed resource depending on the validity of the lease periods in both locations. The gardener reconciliation can update the lease status on the local resource in the namespace of the co-ordinator.
+
+The remote/global/distributed source for lease status is still required to handle the scenario where the gardener is not able to access the shoot control-plane on the seed cluster.
+
+The assumption is that the opposite melt-down scenario of co-ordinator scaling up the control-plane components while gardener reconciliation scales it down will not happen because in such a scenario the gardener reconciliation would actually delete the shoot control-plane rather than just scale down.
+
+### Pros
+
+* No changes to existing shoot control-plane components except for additional livenessProbes.
+* Works for different mechanisms for advertising the current seed cluster lease for a shoot cluster.
+
+### Cons
+
+* One more additional component (shoot control-plane co-ordinator) in the shoot control-plane. Gardener reconciliation should be enhanced to handle the shoot control-plane co-ordinator.
+* Needs some sort of mechanism to advertise current seed cluster lease for a shoot cluster which need to be actively or passively kept up-to-date with changes to the shoot YAML in the garden cluster.
+
+### Changes Required in Gardener Reconciliation
+
+On movement of seed for a shoot control-plane,
+
+1. Update the lease status (this is optional if the lease status is calculated based on shoot YAML).
+2. Deploy the new control-plane on the destination seed cluster. This can done even before the current lease expires.
+3. Garbage-collect the old shoot control-plane on the source seed cluster. This should be done after the new shoot control-plane is available and ready on the destination seed cluster.
+   * Some care would be required while deleting the old shoot control-plane on the source seed cluster to avoid deleting the actual backup object store or even trigger the deletion grace period for it. Only the backup resources should be deleted without actually deleting the object store.
+
+## Alternatives
+
+Let individual components in the shoot control-plane (machine-controller-manager, etcd statefulset, VPN etc.) manage lease and co-ordination themselves.
+
+This approach was discarded because of unnecessary duplication of responsibility.
```

Comment:
I would also believe, a central coordinator controlling the control plane, makes more sense.

Reply:
This is why a livenessProbe is required in MCM checking on the health of the co-ordinator. The same is required in etcd statefulset as well because the old etcd backup sidecar should stop taking backups after migration regardless of whether the co-ordinator is healthy and able to scale down etcd or not.

Reply:
The livenessProbe doesn't solve the problem in a safe way, does it? The component may be stopped too late. It doesn't happen immediately and I was merely hinting at the theoretical problem and inherently unsafe synchronisation means.